pkg/iam_roles/iam_role.go

```diff
@@ -0,0 +1,134 @@
+	"k8s.io/apimachinery/pkg/types"
+	"regexp"
+	controllerruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// IamRoleConfig : config for AWS IAM role settings for a paasta container
+type IamRoleConfig struct {
+	// +optional
+	IamRoleProvider *string `json:"iam_role_provider,omitempty"`
+	// +optional
+	IamRole *string `json:"iam_role,omitempty"`
+	// PAASTA-16919: remove everything related to fs_group when
+	// https://github.com/aws/amazon-eks-pod-identity-webhook/issues/8 will be
+	// fixed.
+	// +optional
+	FsGroup *int64 `json:"fs_group,omitempty"`
+}
+
+var defaultIamRoleProvider = "kiam"
+var DefaultFsGroup int64 = 65534
+
+var serviceAccountRegex = regexp.MustCompile("[^0-9a-zA-Z]+")
+
+// SetIamRoleConfigDefaults: sets the default values for the AWS IAM role config
+func SetIamRoleConfigDefaults(iamRoleConfig *IamRoleConfig) {
+	if iamRoleConfig.IamRoleProvider == nil {
+		iamRoleConfig.IamRoleProvider = &defaultIamRoleProvider
+	}
+	if iamRoleConfig.FsGroup == nil {
+		iamRoleConfig.FsGroup = &DefaultFsGroup
+	}
+}
+
+// EnsureForIamRole: prepare AWS IAM fole for use
+func EnsureForIamRole(ctx context.Context, client controllerruntimeclient.Client, namespace string, iamRoleConfig *IamRoleConfig) error {
+	// We need to create a service account only for "aws" provider
+	if iamRoleConfig.IamRoleProvider != nil && *iamRoleConfig.IamRoleProvider == "aws" {
+		glog.V(4).Infof("Ensuring service account in %s for iam_role %v exists", namespace, iamRoleConfig.IamRole)
+		if iamRoleConfig.IamRole == nil {
+			return fmt.Errorf("%s/%v: iam_role must be specified when iam_role_provider is set to 'aws'", namespace, iamRoleConfig.IamRole)
+		}
+
+		saName := getServiceAccountNameForIamRole(iamRoleConfig.IamRole)
+
+		// check if service account with this name+namespace already exists
+		glog.V(4).Infof("Looking for service account called %s in namespace %s", saName, namespace)
+		result := &corev1.ServiceAccount{}
+		err := client.Get(ctx,
+			types.NamespacedName{
+				Name:      saName,
+				Namespace: namespace,
+			},
+			result,
+		)
+
+		if err != nil {
+			glog.Errorf("Error while getting service account: %s", err)
+			if errors.IsNotFound(err) {
+				glog.Infof("Service account not found, creating it\n")
+				annotations := map[string]string{
+					"eks.amazonaws.com/role-arn": *iamRoleConfig.IamRole,
+				}
+				service_account := &corev1.ServiceAccount{
+					TypeMeta: v1.TypeMeta{},
+					ObjectMeta: v1.ObjectMeta{
+						Name:        saName,
+						Namespace:   namespace,
+						Annotations: annotations,
+					},
+					Secrets:                      nil,
+					ImagePullSecrets:             nil,
+					AutomountServiceAccountToken: nil,
+				}
+				err = client.Create(ctx, service_account)
+				if err != nil {
+					return err
+				}
+				return nil
+			}
+			// some other error occurred
+			return err
+		}
+
+		// service account already exists so do nothing
+		glog.V(4).Info("Service account already exists")
```

Comment:
the thing I was getting at in the other review was that we probably want a way to update fields other than just the name + namespace. f.e, if we update the annotations on the service account one day, then the sa will never be updated.

to solve it, I think we probably want to hash the target object + existing object (like we do elsewhere) and delete + recreate (or update) if it changes.

Reply:
I really prefer to keep changes in this PR as simple and "copied from Flink operator" as possible.
Let's do the hash-related changes in another one PR.